src/c/device-coap.c

```diff
@@ -0,0 +1,225 @@
+  }
+
+  /* CoAP server bind address as text */
+  const char *bind_addr = iot_data_string_map_get_string (config, COAP_BIND_ADDR_KEY);
+  if (bind_addr)
+  {
+    driver->coap_bind_addr = iot_data_alloc_string (bind_addr, IOT_DATA_COPY);
+  }
+  else
+  {
+    iot_log_error (lc, "CoAP bind address not in configuration");
+    driver->coap_bind_addr = NULL;
+    return false;
+  }
+
+  iot_log_debug (lc, "Init complete");
+  return true;
+}
+
+static bool coap_get_handler
+(
+  void *impl,
+  const char *devname,
+  const devsdk_protocols *protocols,
+  uint32_t nreadings,
+  const devsdk_commandrequest *requests,
+  devsdk_commandresult *readings,
+  const devsdk_nvpairs *qparms,
+  iot_data_t **exception
+)
+{
+  (void) impl;
+  (void) devname;
+  (void) protocols;
+  (void) nreadings;
+  (void) requests;
+  (void) readings;
+  (void) qparms;
+  (void) exception;
+  return true;
+}
+
+static bool coap_put_handler
+(
+  void *impl,
+  const char *devname,
+  const devsdk_protocols *protocols,
+  uint32_t nvalues,
+  const devsdk_commandrequest *requests,
+  const iot_data_t *values[],
+  iot_data_t **exception
+)
+{
+  (void) impl;
+  (void) devname;
+  (void) protocols;
+  (void) nvalues;
+  (void) requests;
+  (void) values;
+  (void) exception;
+  return true;
+}
+
+static void coap_stop (void *impl, bool force) {}
+
+int main (int argc, char *argv[])
+{
+  coap_driver * impl = malloc (sizeof (coap_driver));
+  memset (impl, 0, sizeof (coap_driver));
+
+  devsdk_error e;
+  e.code = 0;
+
+  /* Device Callbacks */
+  devsdk_callbacks coapImpls =
+  {
+    coap_init,
+    NULL,              /* discovery */
+    coap_get_handler,
+    coap_put_handler,
+    coap_stop,
+    NULL,              /* auto-event started */
+    NULL,              /* auto-event stopped */
+    NULL,              /* device added */
+    NULL,              /* device updated */
+    NULL               /* device removed */
+  };
+
+  /* Initialize a new device service */
+  devsdk_service_t *service = devsdk_service_new
+    ("device-coap", VERSION, impl, coapImpls, &argc, argv, &e);
+  ERR_CHECK (e);
+  impl->service = service;
+
+  int res = 0;
+  int n = 1;
+  while (n < argc)
+  {
+    if (strcmp (argv[n], "-h") == 0 || strcmp (argv[n], "--help") == 0)
+    {
+      printf ("Usage: device-coap [options]\n");
+      printf ("\nOptions:\n");
+      printf ("  -r, --registry  set registry URL\n");
+      printf ("  -i, --instance  set instance name\n");
+      printf ("  -p, --profile   set profile file\n");
+      printf ("  -c, --confdir   set configuration directory\n");
+      printf ("  -f, --file      set configuration file\n");
+      printf ("  -h, --help      show this text\n");
+      goto end;
+    }
+    else
+    {
+      printf ("%s: Unrecognized option %s\n", argv[0], argv[n]);
+      res = 1;
+      goto end;
+    }
+  }
+
+  devsdk_service_start (service, &e);
+  ERR_CHECK (e);
+
+  /* Validate transport security mode from config/environment */
```

Comment:
Should make this an error condition in coap_init

Reply:
Addressed in 5b008dc